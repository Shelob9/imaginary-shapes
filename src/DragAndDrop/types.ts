export type dragAndDropItem = {
	id: string;
	content: string;
};
export type dragAndDropItems = { [key: string]: dragAndDropItem };

export type dragAndDropColumm = {
	id: string;
	title: string;
	itemIds: Array<string>;
};
export type dragAndDropState = {
	items: dragAndDropItems;
	columns: { [key: string]: dragAndDropColumm };
	columnOrder: Array<string>;
};
