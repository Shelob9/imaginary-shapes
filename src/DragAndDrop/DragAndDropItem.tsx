import React from "react";
import { Draggable } from "react-beautiful-dnd";
import { dragAndDropItem } from "./types";

export default class DragAndDropItem extends React.Component<{
	item: dragAndDropItem;
	index: number;
}> {
	render() {
		const isDragDisabled = this.props.item.id === "task-1";
		return (
			<Draggable
				draggableId={this.props.item.id}
				index={this.props.index}
				isDragDisabled={isDragDisabled}
			>
				{(provided, snapshot) => (
					<div
						ref={provided.innerRef}
						{...provided.draggableProps}
						{...provided.dragHandleProps}
						innerRef={provided.innerRef}
						isDragging={snapshot.isDragging}
						isDragDisabled={isDragDisabled}
					>
						{this.props.item.content}
					</div>
				)}
			</Draggable>
		);
	}
}
