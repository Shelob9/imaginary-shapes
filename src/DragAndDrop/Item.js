import React from 'react'
import { Draggable } from 'react-beautiful-dnd'

export default class Item extends React.Component {
  render() {
    const isDragDisabled = this.props.task.id === 'task-1'
    return (
      <Draggable
        draggableId={this.props.task.id}
        index={this.props.index}
        isDragDisabled={isDragDisabled}
      >
        {(provided, snapshot) => (
          <div
            ref={provided.innerRef}
            {...provided.draggableProps}
            {...provided.dragHandleProps}
            innerRef={provided.innerRef}
            isDragging={snapshot.isDragging}
            isDragDisabled={isDragDisabled}
          >
            {this.props.task.content}
          </div>
        )}
      </Draggable>
    )
  }
}
