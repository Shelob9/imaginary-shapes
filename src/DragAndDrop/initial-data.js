const initialData = {
  items: {
    'task-1': { id: 'task-1', content: 'Take out the garbage' },
    'task-2': { id: 'task-2', content: 'Watch my favorite show' },
    'task-3': { id: 'task-3', content: 'Charge my phone' },
    'task-4': { id: 'task-4', content: 'Cook dinner' }
  },
  columns: {
    topLeft: {
      id: 'topLeft',
      title: 'Top Left',
      itemIds: ['task-1', 'task-2', 'task-3', 'task-4']
    },
    topRight: {
      id: 'topRight',
      title: 'Top Right',
      itemIds: []
    },
    bottomLeft: {
      id: 'bottomLeft',
      title: 'Bottom Left',
      itemIds: []
    },
    bottomRight: {
      id: 'bottomRight',
      title: 'Bottom Right',
      itemIds: []
    }
  },
  // Facilitate reordering of the columns
  columnOrder: ['topLeft', 'topRight', 'bottomLeft', 'bottomRight']
}

export default initialData
