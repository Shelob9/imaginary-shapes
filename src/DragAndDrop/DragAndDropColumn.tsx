import React from "react";
import DragAndDropItem from "./DragAndDropItem";
import { Droppable } from "react-beautiful-dnd";
import { dragAndDropColumm, dragAndDropItem } from "./types";
class List extends React.Component<{ innerRef: React.Ref<any> }> {
	render() {
		return <div ref={this.props.innerRef}>{this.props.children}</div>;
	}
}

export default class DragAndDropColumn extends React.Component<{
	column: dragAndDropColumm;
	items: Array<dragAndDropItem>;
}> {
	render() {
		return (
			<div>
				<Droppable droppableId={this.props.column.id} type="TASK">
					{(provided, snapshot) => (
						<List
							innerRef={provided.innerRef}
							{...provided.droppableProps}
							isDraggingOver={snapshot.isDraggingOver}
						>
							{this.props.items.map((item: dragAndDropItem, index) => (
								<DragAndDropItem key={item.id} item={item} index={index} />
							))}
							{provided.placeholder}
						</List>
					)}
				</Droppable>
			</div>
		);
	}
}
