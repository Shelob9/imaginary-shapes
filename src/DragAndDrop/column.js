import React from 'react'
import Item from './Item'
import { Droppable } from 'react-beautiful-dnd'

class Container extends React.Component {
  render() {
    return <div ref={this.props.innerRef}>{this.props.children}</div>
  }
}

class TaskList extends React.Component {
  render() {
    return <div ref={this.props.innerRef}>{this.props.children}</div>
  }
}

export default class Column extends React.Component {
  render() {
    return (
      <Container>
        <h3>{this.props.column.title}</h3>
        <Droppable droppableId={this.props.column.id} type="TASK">
          {(provided, snapshot) => (
            <TaskList
              innerRef={provided.innerRef}
              {...provided.droppableProps}
              isDraggingOver={snapshot.isDraggingOver}
            >
              {this.props.items.map((task, index) => (
                <Item key={task.id} task={task} index={index} />
              ))}
              {provided.placeholder}
            </TaskList>
          )}
        </Droppable>
      </Container>
    )
  }
}
