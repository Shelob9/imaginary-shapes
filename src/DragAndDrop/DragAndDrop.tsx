import React from "react";
import { DragDropContext } from "react-beautiful-dnd";

import DragAndDropColumn from "./DragAndDropColumn";
import { dragAndDropState } from "./types";

class DragAndDrop extends React.Component<
	{
		initialData: dragAndDropState;
		//Optionally call a function to modify state before updating
		stateMiddleWare?: (update: dragAndDropState) => dragAndDropState;
	},
	dragAndDropState
> {
	constructor(props) {
		super(props);
		this.state = props.initialData;
	}

	//IMPORTANT: ALWAYS use this to set state
	updateState = (update: dragAndDropState) => {
		if (this.props.stateMiddleWare) {
			this.setState(this.props.stateMiddleWare(update));
		} else {
			this.setState(update);
		}
	};

	onDragEnd = (result) => {
		const { destination, source, draggableId } = result;

		if (!destination) {
			return;
		}

		if (
			destination.droppableId === source.droppableId &&
			destination.index === source.index
		) {
			return;
		}

		const start = this.state.columns[source.droppableId];
		const finish = this.state.columns[destination.droppableId];

		if (start === finish) {
			const newitemIds = Array.from(start.itemIds);
			newitemIds.splice(source.index, 1);
			newitemIds.splice(destination.index, 0, draggableId);

			const newColumn = {
				...start,
				itemIds: newitemIds,
			};

			const newState = {
				...this.state,
				columns: {
					...this.state.columns,
					[newColumn.id]: newColumn,
				},
			};

			this.updateState(newState);
			return;
		}

		// Moving from one list to another
		const startitemIds = Array.from(start.itemIds);
		startitemIds.splice(source.index, 1);
		const newStart = {
			...start,
			itemIds: startitemIds,
		};

		const finishitemIds = Array.from(finish.itemIds);
		finishitemIds.splice(destination.index, 0, draggableId);
		const newFinish = {
			...finish,
			itemIds: finishitemIds,
		};

		const newState = {
			...this.state,
			columns: {
				...this.state.columns,
				[newStart.id]: newStart,
				[newFinish.id]: newFinish,
			},
		};
		this.updateState(newState);
	};

	render() {
		return (
			<DragDropContext onDragEnd={this.onDragEnd}>
				<div>
					{this.state.columnOrder.map((columnId) => {
						const column = this.state.columns[columnId];
						const items = column.itemIds.map(
							(itemId) => this.state.items[itemId]
						);

						return (
							<DragAndDropColumn
								key={column.id}
								column={column}
								items={items}
							/>
						);
					})}
				</div>
			</DragDropContext>
		);
	}
}

export default DragAndDrop;
