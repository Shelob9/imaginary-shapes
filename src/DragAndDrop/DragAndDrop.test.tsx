import React from "react";
import { render, fireEvent } from "@testing-library/react";
import DragAndDrop from "./DragAndDrop";
import initialData from "./initial-data";

describe("DragAndDrop component", () => {
	let props = { initialData };
	it(" DragAndDrop matches snapshot", () => {
		props = { ...props };
		const { container } = render(<DragAndDrop {...props} />);
		expect(container).toMatchSnapshot();
	});
	it.skip("DragAndDrop calls on change", () => {
		const onChange = jest.fn();
		props = { ...props };
		const { getByLabelText } = render(
			<DragAndDrop {...props} onChange={onChange} />
		);
		fireEvent.change(getByLabelText("Label Text"), {
			target: { value: "Stillness" },
		});
		expect(onChange).toBeCalledTimes(1);
		expect(onChange).toBeCalledWith("Stillness");
	});
});
