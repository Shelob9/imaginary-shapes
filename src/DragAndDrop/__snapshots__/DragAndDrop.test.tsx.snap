// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragAndDrop component  DragAndDrop matches snapshot 1`] = `
<div>
  <div>
    <div>
      <h3>
        Top Left
      </h3>
      <div>
        <div
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="task-1"
        >
          Take out the garbage
        </div>
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="task-2"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="task-2"
          draggable="false"
          role="button"
          tabindex="0"
        >
          Watch my favorite show
        </div>
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="task-3"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="task-3"
          draggable="false"
          role="button"
          tabindex="0"
        >
          Charge my phone
        </div>
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="task-4"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="task-4"
          draggable="false"
          role="button"
          tabindex="0"
        >
          Cook dinner
        </div>
      </div>
    </div>
    <div>
      <h3>
        Top Right
      </h3>
      <div />
    </div>
    <div>
      <h3>
        Bottom Left
      </h3>
      <div />
    </div>
    <div>
      <h3>
        Bottom Right
      </h3>
      <div />
    </div>
  </div>
</div>
`;
